inst/include/sampler.h

```diff
@@ -246,6 +247,75 @@ namespace smc {
         delete pAdapt;
     }
 
+    template <class Space, class Params>
+    void sampler<Space, Params>::_copy(const sampler<Space,Params> & sFrom)
+    {
+        ///Number of particles in the system.
+        N = sFrom.N;
+        ///The current evolution time of the system.
+        T = sFrom.T;
+
+        ///The resampling mode which is to be employed.
+        rtResampleMode = sFrom.rtResampleMode;
+        ///The effective sample size at which resampling should be used.
+        dResampleThreshold = sFrom.dResampleThreshold;
+
+        ///Structure used internally for resampling.
+        dRSWeights = sFrom.dRSWeights;
+        ///Structure used internally for resampling.
+        uRSCount = sFrom.uRSCount;
+        ///Structure used internally for resampling.
+        uRSIndices = sFrom.uRSIndices;
+
+        ///The particles within the system.
+        pPopulation = sFrom.pPopulation;
+        ///The set of moves available.
+        Moves = sFrom.Moves;
+        /// The additional algorithm parameters.
```

Comment:
Sorry for being rather late with this, I've been swamped with other things.

Thanks for this, Binxinag, it's always good to sort out these small issues. 

However, there is a bit of a complication, at this stage the code copies a pointer and also an indicator of whether the object pointed to belongs to a sampler. If you copy a sampler which owns pAdapt and delete both of them then they will both delete the same object.

I suspect that the cleanest option would be to do a deep copy of the object pointed to and then to set adaptBelong to TRUE regardless of what the original configuration was, but I haven't thought very much about this.

Any other thoughts on this?

Reply:
Got it. Perhaps we could do it like 
```
 *pAdapt = *sFrom.pAdapt;
 adaptBelong = sFrom.adaptBelong;
```
Then even if `adaptBelong == true`, we do not delete the same object.

Reply:
Then one new problem may come out. When pAdapt is null at first, *pAdapt may bring segment fault. :(

Reply:
Quite so. This is where pAdapt would be initialised so to go down that route you'd need to allocate memory before assigning to it -- but that wouldn't need much more than an additional 
`pAdapt = new...`
assuming that the deep copy model is the approach we want to go with here.

Also, if we do do this then it should probably be 
`adaptBelong = true`
rather than 
`adaptBelong = sFrom.adaptBelong` 
because even if the source object doesn't own the object pointed to by its `pAdapt` the copied object does.

Reply:
Thanks Adam for picking up on this. I'd forgotten about this issue of dealing with the pointer in the copy. I think I put this off because I didn't have a great solution at the time but I like the sound of your solution.

Reply:
There's been a history of omitting the copy constructor because it was fiddly for various reasons dating right back to the first implementation. I think we're closing in on something that will at least work -- it will be great to have something in place, but we should remember to be clear in any documentation exactly what the behaviour on copying is.

(The only obvious alternative that I can see would be to always copy the pointer and then set `adaptBelong = false` but that runs into problems if the original sampler owned the object and is destroyed before the copy... but if anyone has other thoughts they'd be very welcome.)